include/lbann/optimizers/optimizer.hpp

```diff
@@ -152,43 +153,40 @@ class optimizer {
   /** Gradient matrix. */
   AbsDistMat* m_gradient;
 #ifdef LBANN_HAS_CUDNN
-  /** GPU memory for gradient matrix. */
-  std::vector<DataType*> m_gradient_d;
+  /** GPU gradient matrix. */
+  cudnn::matrix m_gradient_d;
 #endif // LBANN_HAS_CUDNN
 
  private:
 
-  /** Whether the CPU gradient matrix is non-zero. */
-  bool m_cpu_gradient_is_nonzero;
-  /** Whether the CPU staging matrix is non-zero. */
-  bool m_cpu_staging_is_nonzero;
```

Comment:
In the old implementation, we set these flags to avoid unnecessary CPU-GPU memory transfers. However, it's easier if the CPU implementation always keeps data on CPU and the GPU implementation on GPU.